For udpu, don't check join node's ip was in the same network

liangxin1300 opened the pull request:
## Problem
While in node join process in udpu mode, we do have restriction that forcing the join node address being the same network with the init node's. Corosync doesn't have such limit, so we should loose this checking.

But, loosing this check may have side effect: for normal/non-cloud environment, corosync might be configured with **crossed** link, in which scenario each node can not see each other after joining.

Take an example for **crossed** link:
```
node1's interface: eth1, eth2
node2's interface: eth1, eth2
node1: use eth1 for corosync
node2: use eth2 for corosync
```

So we should have to check the cluster status during joining. 
**If join node itself is online while can't see init node**, bootstrap on join node will:
1. recover original corosync.conf and sync it to init node
2. stop corosync.service on join node
3. exit with an error message 

On join node, it will successfully joined in cluster until using the correct link

## Unit test
a170cb839cce7fb5539e639603058e35d6b57389

## Functional test
b42cbe8eb6bb5070362e3b1fb9dec786fef7e548

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -228,13 +231,49 @@ def wait_for_cluster():
+    peer_node = None
+    if _context.cluster_node:
+        if utils.valid_ip_addr(_context.cluster_node):
+            peer_node = utils.get_stdout("ssh {} hostname".format(_context.cluster_node))[1]
```
What would happen here in this scenario:
- The IP address is valid, but it's not available for SSH or it doesn't even exist.

liangxin1300 replied:
Changed, use `utils.get_stdout_stderr` instead, and check the return code

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -1846,6 +1885,9 @@ def update_nodeid(nodeid, node=None):
         else:
             corosync.set_value("totem.nodeid", nodeid)
 
+    atexit.register(os.remove, COROSYNC_CONF_ORIG)
```
What this `register` is doing?

liangxin1300 replied:
run `os.remove(COROSYNC_CONF_ORIG)` when the program terminated

arbulu89 replied:
Ok, thanks!

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -228,13 +231,49 @@ def wait_for_cluster():
+    if not re.search("Online: .* {} ".format(utils.this_node()), crm_mon_txt):
+        return False
+
+    peer_node = get_peer_hostname()
```
I would directly return True here if peer_node is False. It would make the code easier to read.
And we could add comments.
If peer_node is False, it means that this is a `init` process, right? I would add this comments there

liangxin1300 replied:
OK, changed

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -1909,21 +1951,6 @@ def update_nodeid(nodeid, node=None):
                                               ringXaddr_res)
                 if not ringXaddr:
                     error("No value for ring{}".format(i))
-
```
Is this code executed if unicast is set? I cannot really understand the whole process.
I'm just afraid if this can affect to multicast

arbulu89 replied:
In fact, why do we have things related to multicast in that method?

liangxin1300 replied:
This block of code is under flag `is_unicast`, if this flag is Flase(which means is in multicast mode), this code wouldn't be executed.
So this block of code has no related with multicast

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -228,13 +231,49 @@ def wait_for_cluster():
+    Check whether local node is online
+    Besides that, in join process, check whether init node is online
+    """
+    if not re.search("Online: .* {} ".format(utils.this_node()), crm_mon_txt):
```
`is_online` is called by `wait_for_cluster`, which is just after running `start_service("pacemaker.service")`, while `start_service` will check whether `pacemaker.service` is successfully started, if not, the whole program will return with an error.

So, generally, the `this_node` will always online, just wait for some seconds.

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -228,13 +231,49 @@ def wait_for_cluster():
+
+    peer_node = get_peer_hostname()
+    # In join process
+    # If the join node has already online but can't see the init node
```
Changed

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -228,13 +231,49 @@ def wait_for_cluster():
+        csync2_update(corosync.conf())
+        stop_service("corosync")
+        print()
+        error("Can not see peer node \"{}\", please check the communicate IP".format(peer_node))
```
changed

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -228,13 +231,49 @@ def wait_for_cluster():
     status_done()
 
 
+def get_peer_hostname():
```
changed

liangxin1300 commented:
> Hi @liangxin1300 ,
> I cannot really say if the changes are good or not yet as I don't understand many things.
> Just to clarify some doubts:
> When unicast is used, do all the nodes need to use the same network interface?
> Is this PR affecting somehow to multicast usage?

@arbulu89 , thanks for your suggestions, I have changed and updated.

Besides that, I notice that there was also another side effect for loosing this limit, like this scenarios below:
```
on joining node:
eth0's ip already successfully configured in corosync.conf

then run crm cluster stop
then run crm cluster join -c init_node -i eth2 -y
then nodelist has three node sections:
  1. init node's ip
  2. join node eth0's ip
  3. join node eth1's ip
```
Above issue is similar with bsc#1127095,
So I append commit c2919536bb1145cc3e5115e60b923706c4393c46 and update related unittest code to resolve above issue

arbulu89 commented on crmsh/corosync.py:
```diff
@@ -811,20 +811,48 @@ class IPAlreadyConfiguredError(Exception):
-    p = Parser(f)
-
-    # to check if the same IP already configured
+def find_configured_ip(IParray):
```
What is this IParray?
Is it a list with ip addresses? If this is a variable, we should change `IParray` if `iparray`. Otherwise, it looks like a constant, and I don't know if this is the case. It would make the code more readble

liangxin1300 replied:
Changed as `ip_list`

arbulu89 commented on crmsh/corosync.py:
```diff
@@ -811,20 +811,48 @@ class IPAlreadyConfiguredError(Exception):
+        p = Parser(f.read())
+
+    # get exist ip list from corosync.conf
     exist_iplist = []
```
I would change this variable names. `exist_iplist` by `corosync_ip_list`

liangxin1300 replied:
changed

arbulu89 commented on crmsh/corosync.py:
```diff
@@ -811,20 +811,48 @@ class IPAlreadyConfiguredError(Exception):
+    else:
+        all_possible_ip = IParray
+
+    for ip in set(all_possible_ip):
```
I think you can change this for loops by (if both are `sets`). This checks if some value is in both lists:
```
if bool(all_possible_ip & exist_iplist):
    raise IPAlreadyConfiguredError("IP {} was already configured".format(ip))
```

arbulu89 commented on crmsh/corosync.py:
```diff
@@ -811,20 +811,48 @@ class IPAlreadyConfiguredError(Exception):
+    #   eth1's ip also want to add in nodelist
+    # if this scenarios happened, raise IPAlreadyConfiguredError
     for ip in IParray:
+        if ip in local_ip_list:
```
I don't understand much this piece of code.
Here you check if some local IP is in the `IParray` list and if just 1 is there, you add the whole list.
Why do you add the whole list? Shouldn't you add just the found IP address?

By the way, you can replace this for and if by:
```
if bool(IParray & local_ip_list):
```

liangxin1300 replied:
Consider this scenario:
Join node eth0's ip `10.10.10.2` already configured in corosync.conf;
then join node eth1's ip `20.20.20.2` want to add in corosync.conf;
Local ip list in join node is `["10.10.10.2", "20.20.20.2"]`,
if we don't extend local ip list to `all_possible_ip`, no eception `IPAlreadyConfiguredError` will raise, `20.20.20.2` will add to corosync.conf

arbulu89 replied:
But at the end you are adding all the IP addresses in `ip_list` to `all_possible_ip`, even though only one might be in both lists. For example:
```
ip_list = [1.1.1.1, 2.2.2.2]
local_ip_list = [1.1.1.1, 3.3.3.3, 4.4.4.4]
all_possible_ip = [1.1.1.1, 2.2.2.2, 1.1.1.1, 3.3.3.3, 4.4.4.4]
```
Only `1.1.1.1` matches but you add all of the addresses. Is this what you want? Just to understand better.

If this is what you want, and to make the code easier. I would just to this:
```
all_ip_list = set(ip_list)
if (set(ip_list) & set(local_ip_list)):
    all_ip_list |= set(local_ip_list)
if bool(all_possible_ip & set(corosync_ip_list)):
    raise IPAlreadyConfiguredError("IP {} was already configured".format(ip))
```
This piece of code does:
1. If some value in ip_list is in local_ip list -> all_ip_list = ip_list + local_ip_list, else all_ip_list = ip_list
2. If some value in all_ip_list is in corosync_ip_list, raise the error

liangxin1300 replied:
Changed.
Thank you @arbulu89

gao-yan replied:
Does it mean we don't mind if any IPs from ip_list are not in local_ip_list? We'd just try whatever the user is asking for? Not really a problem since there's fallback.  Just wondering :-)

arbulu89 commented on crmsh/corosync.py:
```diff
@@ -811,20 +811,48 @@ class IPAlreadyConfiguredError(Exception):
             raise IPAlreadyConfiguredError("IP {} was already configured".format(ip))
 
+
+def add_node_ucast(IParray, node_id=None):
```
The same than before, if this is a variable and not a constant, change `IParray` by `iparray`

liangxin1300 replied:
Changed as `ip_list`

gao-yan commented on crmsh/bootstrap.py:
```diff
@@ -1846,6 +1889,9 @@ def update_nodeid(nodeid, node=None):
             corosync.set_value("totem.nodeid", nodeid)
 
+    atexit.register(os.remove, COROSYNC_CONF_ORIG)
+    shutil.copy(corosync.conf(), COROSYNC_CONF_ORIG)
```
Would it be better to use `tempfile.mktemp()` to create the temporary file, just in case this would accidentally override any existing files of users ? AFAICT, crmsh uses tempfile to create temporary things for other purposes as well.

liangxin1300 replied:
Changed. Use `tmpfiles.create` to create temporary file.
Thanks!

liangxin1300 replied:
Already changed for crmsh-4.1 and crmsh-3.0 branche PRs

gao-yan commented on crmsh/bootstrap.py:
```diff
@@ -228,13 +231,53 @@ def wait_for_cluster():
+    peer_node = None
+    if _context.cluster_node:
+        if utils.valid_ip_addr(_context.cluster_node):
+            rc, out, err = utils.get_stdout_stderr("ssh {} hostname".format(_context.cluster_node))
```
Since a system might have multiple names corresponding to different IP addresses in /etc/hosts, to make sure we are really getting the `node name` used for the cluster, we'd better run `crm_node --name` on that node but of course while cluster service is running there.

gao-yan replied:
And of course, probably depending on how this function will be potentially used, use `uname --n` as a fallback in case `crm_node --name` fails for example if cluster service is not running there yet?

liangxin1300 replied:
Changed as `crm_node --name`, and also changed for crmsh-4.1 and crmsh-3.0 branche PRs.

I think if the cluster service is really stopped on init node while executing this piece of code, just let `error` to terminate the join process, because in that case, we should stop and check why service on init node stopped, instead of going ahead.

Thanks @gao-yan !
